torch/autograd/gradcheck.py

```diff
@@ -661,6 +739,208 @@ def transpose(matrix_of_tensors):
+    # Compute jacobians in slow mode for better error message
+    slow_numerical = _get_numerical_jacobian(func, tupled_inputs, outputs)[input_idx][output_idx]
+    slow_analytical = _get_analytical_jacobian(tupled_inputs, outputs, input_idx, output_idx)
+
+    # Assume jacobians are non-empty and have the same shape
+    slow_max_diff = (slow_numerical - slow_analytical).abs().max()
+
+    slow_allclose = torch.allclose(slow_analytical, slow_numerical, rtol, atol)
+    msg = ("\nThe above quantities relating the numerical and analytical jacobians are computed \n"
+           "in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background \n"
+           "about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:\n\n"
+           f"Numerical:\n {slow_numerical}\n"
+           f"Analytical:\n{slow_analytical}\n\n"
+           f"The max per-element difference (slow mode) is: {slow_max_diff}.\n")
+    if slow_allclose:
+        # Slow gradcheck would've passed!
+        msg += FAST_FAIL_SLOW_OK_MSG
+    return msg
+
+
+def fast_gradcheck(fail_test, func, func_out, tupled_inputs, outputs, eps, rtol,
+                   atol, raise_exception, check_grad_dtypes, nondet_tol):
+    # Perform the fast version of gradcheck
+    # See https://github.com/pytorch/pytorch/issues/53876 for details
+    inp_tensors = [t for t in tupled_inputs if is_tensor_like(t) and t.requires_grad]
+    inp_tensor_indices = [i for i, t in enumerate(tupled_inputs) if is_tensor_like(t) and t.requires_grad]
+
+    g_cpu = torch.Generator()
+    all_u = [vec_from_tensor(inp, g_cpu) for inp in inp_tensors]
+    all_u_dense = [u.to_dense().reshape(-1) if u.layout == torch.sparse_coo else u for u in all_u]
+    all_v = [vec_from_tensor(out, g_cpu) for out in outputs]
+
+    if not outputs:
+        if not check_no_differentiable_outputs_fast(fail_test, func, func_out, tupled_inputs, inp_tensor_indices,
+                                                    all_u, eps, nondet_tol):
+            return False
+
+    any_complex = any(o.is_complex() for o in outputs)
+    complex_output_indices = [i for i, o in enumerate(outputs) if o.is_complex()]
+
+    # Initialize list of lists to store jacobians for each input, output pair
+    all_analytical: List[List[torch.Tensor]] = [[] for _ in outputs]
+    all_numerical: List[List[torch.Tensor]] = [[] for _ in inp_tensors]
+    all_analytical_from_imag_grad_out: List[List[torch.Tensor]] = [[] for _ in complex_output_indices]
+    all_numerical_from_imag_grad_out: List[List[torch.Tensor]] = [[] for _ in inp_tensors]
+
+    # Numerically approximate v^T (J u)
+    for i, (input_idx, u) in enumerate(zip(inp_tensor_indices, all_u)):
+        numerical = get_fast_numerical_jacobian_wrt_specific_input(func, input_idx, tupled_inputs,
+                                                                   outputs, u, eps, 1.0)
```

Comment:
The name of this function seems weird: `numerical` is technically a `J u, which is not a jacobian, right? It's more of a "jup" (jacobian-u-product)

Reply:
Updated to: get_jvp_wrt_specific_input